### PR TITLE
Adding correlation IDs to reconcile loops

### LINKED ISSUE #1310: Azure API requests through the SDK should use a shared correlation ID for a given reconcile loop
/kind feature

**Describe the solution you'd like**
A single reconcile loop is a correlated event. The API requests for that reconcile loop should be correlated with a unique identifier so that it would be easy for an Azure customer support rep to understand related API activity if a problem should occur for a Cluster API Provider Azure user. Azure APIs support correlation IDs as described in the [common API details doc](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#client-request-headers).

To enable this, we should add a correlation ID at the beginning of each reconcile loop by creating a GUID and adding it as a value on the `context.Context`. That context will flow down through each service and API call. On each API request we should ensure that we look in the context for the value and set it the value for the `x-ms-correlation-request-id` header.

To validate the header is properly set, the implementer should work with either @devigned or @CecileRobertMichon to verify the correlation ID is propagated to Azure.

### PR DESCRIPTION
**Pull request by arschles**
Signed-off-by: Aaron Schlesinger <aaron@ecomaz.net>

 

 
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

This PR adds shared correlation IDs to `context.Context`s passed to reconcile loops, so that the API requests made to Azure get logged and tracked properly. That feature is especially useful for debugging.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1310

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Add 'x-ms-correlation-id' headers to all Azure API calls via distributed traces.
```

### CODE REVIEW

**Comment by CecileRobertMichon** [comment]
/ok-to-test

Welcome @arschles!

**Review by devigned — COMMENTED** [review]
WDYT about changing the way that `tele.Tracer.Start()` works such that anytime start is called and the context does not contain a correlationID, one is generated and added to the value bag. This could be done as a best effort kind of thing, so the error could be ignored, or a default value substituted if a new UUID can not be created?

Also, `pkg/trace/corr.go` requires a file header comment. You will see them littered through the other code files. If you run `make test`, it will generate files, lint, and run the local tests.

**Comment by arschles** [comment]
@devigned I think that's a great idea! I'm working on both the `tele.Tracer.Start()` change and fixing build issues now.

File: pkg/trace/span.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+// See below for an example:
+//
+//	ctx := context.Background()
+//	ctx, corrID, newSpan, err := StartSpan(ctx, tele.Tracer(), "myNewSpan")
+//	if err != nil {
+//		log.Fatal(err)
+//	}
+//	defer newSpan.End()
+//	doSomething(ctx)
+func StartSpan(
```
Comment (devigned) [inline pkg/trace/span.go:54]: WDYT of something like this:

In `./util/tele.go`, we wrap the `otel.Tracer` with our own struct that implements `trace.Tracer`. We then write our own `Start(...)` func which would then delegate back to `otel.Tracer().Start(...)`. That way the interface can stay the same.
```go
package tele

type tracer struct {
    otel.Tracer
}

func (t *tracer) Start(ctx, opName string, opt ...otel.SpanOption) (context.Context, otel.Span) {
    // add correlationID to ctx and continue as normal
    // return the inner implementation of Start
    return t.Tracer.Start(ctx, opName, opt...)
}

```

Then either in the same package or elsewhere, add context helpers to set / fetch the correlationID.

```go
package corr  // or something else...

// FromCtx or some better name you think of
func FromCtx(ctx context.Context) (ctx, string) {
    // create new context.WithValue containing correlationID and return both if it does not exist, else return context and correlationID
}

```

Comment (arschles) [reply pkg/trace/span.go:54]: Sure, happy to do that! A few questions before I get started, just to make sure I'm on the same page

- I'm assuming that you'd want to still call `tele.Tracer().Start(...)` in all the reconcilers, correct?
- Are you ok with creating our own tracer instance (the wrapper you mentioned) and then passing that -- instead of the default one -- to `otel.SetTracerProvider` in `main.go`?
- Do we need `FromCtx` to be exported at all? It seems like it would only be used inside the same package as the custom tracer, but not sure if you're thinking something more

Comment (devigned) [reply pkg/trace/span.go:54]: > I'm assuming that you'd want to still call tele.Tracer().Start(...) in all the reconcilers, correct?

If possible, that would be great.

> Are you ok with creating our own tracer instance (the wrapper you mentioned) and then passing that -- instead of the default one -- to otel.SetTracerProvider in main.go?

:100: 

> Do we need FromCtx to be exported at all? It seems like it would only be used inside the same package as the custom tracer, but not sure if you're thinking something more

I was adding that because I saw you were returning the correlationID in your `StartSpan` func. This led me to think you wanted a way to access it from the context. I think one might need to access the correlationID from the context to be able to add that as a HTTP header for the Azure SDK for Go clients.

Comment (arschles) [reply pkg/trace/span.go:54]: perfect - I'll make these changes.

regarding `FromCtx` - just in case someone wants to access the correlation ID, I'll make `FromCtx` exported like you said.

Comment (arschles) [reply pkg/trace/span.go:54]: @devigned just an update, so you know I haven't forgotten about this! I'm looking through the implementation & docs trying to figure out a way to create a new [`TracerProvider`](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#TracerProvider) that we can pass to the `SetTracerProvider` function in `main.go`. That would need to create a new `Tracer` which wraps the implementation you referenced above. I think that's going to require a lot of plumbing, which we could make unnecessary if we called a custom function to get either:

- get a new tracer that which wraps the underlying tracer implementation, like what you showed here
- create a new span, which is what's already in this PR

I'll keep digging, but let me know about going with either of the above options in case there isn't a good way to wrap everything we need

Comment (arschles) [reply pkg/trace/span.go:54]: Another option would be to create a custom `TracerProvider` that wraps the jaeger one. Let me know your thoughts on that too?

Comment (devigned) [reply pkg/trace/span.go:54]: Sorry for the delay, @arschles. I was thinking something more like this: https://github.com/kubernetes-sigs/cluster-api-provider-azure/compare/master...devigned:corr

The above diff should add the correlationID to the context without needing to manipulate the signature of the Tracer. wdyt?

Comment (arschles) [reply pkg/trace/span.go:54]: Hi @devigned I see what you mean now! Thanks, I'll implement that.

**Comment by arschles** [comment]
@devigned I made approximately the change you suggested. the only material difference is that I made the custom `tracer`'s `Start(...)` method create the new `context.Context` with the correlation ID in it, and also call the wrapped `trace.Tracer.Start()` function. let me know if this is more along the same lines as what you're thinking.

File: config/rbac/role.yaml [hunk 2]
```diff
@@ -134,8 +134,6 @@ rules:
   - azuremachinepoolmachines/status
   verbs:
   - get
-  - patch
```
Comment (devigned) [inline config/rbac/role.yaml:137]: I don't think you meant to do this.

File: exp/controllers/azuremachinepool_controller.go [hunk 3]
```diff
@@ -158,19 +157,16 @@ func (ampr *AzureMachinePoolReconciler) SetupWithManager(ctx context.Context, mg
 
 // Reconcile idempotently gets, creates, and updates a machine pool.
 func (ampr *AzureMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
+	ctx, span := tele.Tracer().Start(
```
Comment (devigned) [inline exp/controllers/azuremachinepool_controller.go:160]: I'd hate to lose the trace attributes that were added to this span. Please consider reverting to the previous Start.

File: exp/controllers/azuremanagedcluster_controller.go [hunk 4]
```diff
@@ -100,12 +98,7 @@ func (r *AzureManagedClusterReconciler) Reconcile(ctx context.Context, req ctrl.
 	defer cancel()
 	log := r.Log.WithValues("namespace", req.Namespace, "azureManagedCluster", req.Name)
 
-	ctx, span := tele.Tracer().Start(ctx, "controllers.AzureManagedClusterReconciler.Reconcile",
```
Comment (devigned) [inline exp/controllers/azuremanagedcluster_controller.go:103]: Again, the attributes are valuable. Please consider reverting the change to Start.

**Review by devigned — CHANGES_REQUESTED** [review]
I like the work done in the `tele` package to create the tracer adapter and enhancing the `Start` func to do a bit more.

Please consider reverting the changes to files outside of `./util/tele`. Some of the changes to invocations of the `Start` func changed the attributes being supplied or were solely formatting changes.

File: config/rbac/role.yaml (hunk 2 above)
Comment (arschles) [reply config/rbac/role.yaml]: wowza, the generators might have removed this. sorry about that. re-added in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1460/commits/65ca5042956a85e50911c6a7f22d2975480921f6

File: exp/controllers/azuremanagedcluster_controller.go (hunk 4 above)
Comment (arschles) [reply exp/controllers/azuremanagedcluster_controller.go:103]: sorry about that! re-added in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1460/commits/f21e1c8a0c67c16b223d2d3e3ac57d00706a36a1

File: exp/controllers/azuremachinepool_controller.go (hunk 3 above)
Comment (arschles) [reply exp/controllers/azuremachinepool_controller.go:162]: sorry about that! re-added in https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1460/commits/f21e1c8a0c67c16b223d2d3e3ac57d00706a36a1

**Comment by arschles** [comment]
hey @devigned - would you mind taking another look when you have some time?

**Comment by devigned** [comment]
> Please consider reverting the changes to files outside of `./util/tele`. Some of the changes to invocations of the `Start` func changed the attributes being supplied or were solely formatting changes.

I'm sorry for the delay. I was hoping you were going to revert the changes outside of `./util/tele` since the `Start` interface had stayed the same. If you feel strongly about those formatting changes, then I not opposed. I was just thinking it would make the change set tighter and leave less question about the intent of the changes.

This change set adds the correlationID, but it doesn't actually propagate it on the wire. That will need to be added in a future PR, or this one if you want to continue it.

Again, sorry for any confusion. I'm excited to get this pulled in.

**Comment by CecileRobertMichon** [comment]
@arschles have you thought about how these correlation IDs would get propagated to the actual Azure API calls? I assume a we'd need to set the `x-ms-client-correlation-id` header on each request made to Azure?

**Comment by arschles** [comment]
@devigned got it, I've removed most of the formatting changes made earlier - there are a few still here in which I moved the creation of cancellable contexts below the creation of the spans. that is so the new contexts can be based on the contexts that were created by the new span. I do think it's valuable - but not critical - to do that so the cancellable context has the new correlation IDs in it. let me know what you think of that.

also to address your and @CecileRobertMichon's point about percolating these correlation IDs up to the Azure API - I had originally planned on doing that in a follow-up PR when this PR was large, to draw a clear line between the change to add the IDs and the change to send them up to Azure. now that this is smaller, would you prefer I do that all in this PR?

I'm happy to do it either way, whatever makes it easier for you all

File: exp/controllers/azuremachinepoolmachine_controller.go [hunk 5]
```diff
@@ -127,17 +127,17 @@ func (ampmr *AzureMachinePoolMachineController) SetupWithManager(ctx context.Con
 
 // Reconcile idempotently gets, creates, and updates a machine pool.
 func (ampmr *AzureMachinePoolMachineController) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
-	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultedLoopTimeout(ampmr.ReconcileTimeout))
-	defer cancel()
-	logger := ampmr.Log.WithValues("namespace", req.Namespace, "azureMachinePoolMachine", req.Name)
-
-	ctx, span := tele.Tracer().Start(ctx, "controllers.AzureMachinePoolMachineController.Reconcile",
+	ctx, span := tele.Tracer().Start(ctx, "controllers.AzureMachinePoolController.Reconcile",
```
Comment (CecileRobertMichon) [inline exp/controllers/azuremachinepoolmachine_controller.go:130]:
```suggestion
	ctx, span := tele.Tracer().Start(ctx, "controllers. AzureMachinePoolMachineController.Reconcile",
```

**Comment by CecileRobertMichon** [comment]
> I had originally planned on doing that in a follow-up PR when this PR was large, to draw a clear line between the change to add the IDs and the change to send them up to Azure. now that this is smaller, would you prefer I do that all in this PR?

separate PR is fine with me, up to you

File: exp/controllers/azuremachinepoolmachine_controller.go (hunk 5 above)
Comment (arschles) [reply exp/controllers/azuremachinepoolmachine_controller.go]: Thanks @CecileRobertMichon !

**Review by devigned — APPROVED** [review]
/lgtm
/assign @CecileRobertMichon

**Comment by arschles** [comment]
Thanks @CecileRobertMichon - I'll submit a follow-up PR with those changes

File: exp/controllers/azuremachinepool_controller.go [hunk 6]
```diff
@@ -158,18 +158,18 @@ func (ampr *AzureMachinePoolReconciler) SetupWithManager(ctx context.Context, mg
...
 	ctx, span := tele.Tracer().Start(ctx, "controllers.AzureMachinePoolReconciler.Reconcile",
 		trace.WithAttributes(
 			attribute.String("namespace", req.Namespace),
 			attribute.String("name", req.Name),
 			attribute.String("kind", "AzureMachinePool"),
-		))
+		),
+	)
 	defer span.End()
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultedLoopTimeout(ampr.ReconcileTimeout))
```
Comment (CecileRobertMichon) [inline exp/controllers/azuremachinepool_controller.go:169]: why the change in order here but not in every controller? azuremachine and azurecluster still have the reverse order

Comment (arschles) [reply exp/controllers/azuremachinepool_controller.go:169]: I moved this context creation down to below the creation of the span so that the timeout context includes the correlation ID value. IIRC the others had this change but I reverted them by accident. Would it make more sense to save all these changes for a separate PR?

Comment (CecileRobertMichon) [reply exp/controllers/azuremachinepool_controller.go:169]: either way, as long as the changes are consistent (make the changes here to all the controllers or none of them)

Comment (arschles) [reply exp/controllers/azuremachinepool_controller.go:169]: ok, going to take this change out then because it's not strictly necessary

File: controllers/azuremachine_controller.go [hunk 7]
```diff
@@ -137,7 +137,8 @@ func (r *AzureMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request
 			attribute.String("namespace", req.Namespace),
 			attribute.String("name", req.Name),
 			attribute.String("kind", "AzureMachine"),
-		))
+		),
+	)
```
Comment (arschles) [inline controllers/azuremachine_controller.go:141]: @CecileRobertMichon are you ok with this the the other similar format fixes?
